src/core/qgsrelationmanager.cpp

```diff
@@ -27,8 +27,11 @@ QgsRelationManager::QgsRelationManager( QgsProject *project )
 {
   if ( mProject )
   {
+    // TODO: QGIS 4 remove: relations are now stored with the layer style
```

Comment:
With which style, the referencing or the referenced layer?

Reply:
both: I wanted to be able to reconstruct relations in both ways, but I've only implemented automatic load of referenced layers for now, see the "big question" discussion above: https://github.com/qgis/QGIS/pull/33283#pullrequestreview-328690776

So, for the time being we have some redundancy, but I guess we can live with it, given the potential advantages.

Reply:
Very good points there!

What about a new flag on the relation to make it "soft" or "strong". Where strong relations always try to load the dependency and soft ones are taken as a hint to which appear grayed out first/invalid, and only "fully appear" once both layers are present?

Reply:
I considered the option to implement "weakness" directly inside `QgsRelation`, turning `QgsVectorLayer`s to weak references. 

However having a wrapper looked more clean to me and first of all much simpler, because the current `QgsRelation` (and manager) implementation has a strong precondition on relation validity (if I'm not wrong invalid relations are not added to the manager and removed as soon as the layer becomes invalid).

Reply:
I was more talking about a user configurable option, saved as a boolean in the xml.

Reply:
I don't follow, an option for what?

Reply:
As you write, sometimes, one wants to only work with only a part of a huge data model. And if loading a style, possibly not resolve all relations (and load all layers).

Where to draw the line between "desired" relations (value lists, child classes, etc) and "other, currently uninteresting parts of the model" could be done by tagging some relations as strong (yes, load) and soft (only apply if other layer is already loaded).

But that's not the only way to deal with it, just thinking out loud.

Reply:
Got it, this is actually a good idea, but isn't the "strength" already meant carry that information?

Anyway my "big question" was more about the referenced layers case, I've no doubts that whenever a referencing layer is loaded its referenced counterpart is a strong dependency and must be automatically loaded or the model would be incomplete (and probably some widgets broken).

On the other hand, I'm not sure that we want to automatically load the (potentially many) referencing layers when a referenced layer is loaded.

Anyway all this machinery is controlled by a style that has been saved as a (default) style in the context of a project where relations have been explicitly set by the user, so there is already some control over what gets automatically loaded and what not.

Reply:
Maybe yes. Treat compositions (and aggregations?) as "auto load" and associations as "load on demand" could work. But that's very weak ground I'm basing this assumption on. Reality will probably prove me wrong.